cmd/soroban-rpc/internal/methods/get_events.go

```diff
@@ -341,20 +377,25 @@ func (h eventsRPCHandler) getEvents(request GetEventsRequest) (GetEventsResponse
 		txHash               *xdr.Hash
 	}
 	var found []entry
-	latestLedger, err := h.scanner.Scan(
-		events.Range{
-			Start:      start,
-			ClampStart: false,
-			End:        events.MaxCursor,
-			ClampEnd:   true,
-		},
-		func(event xdr.DiagnosticEvent, cursor events.Cursor, ledgerCloseTimestamp int64, txHash *xdr.Hash) bool {
-			if request.Matches(event) {
-				found = append(found, entry{cursor, ledgerCloseTimestamp, event, txHash})
-			}
-			return uint(len(found)) < limit
-		},
-	)
+	cursorSet := make(map[string]struct{})
+
+	contractIDs := combineContractIDs(request.Filters)
+
+	// Scan function to apply filters
+	f := func(event xdr.DiagnosticEvent, cursor events.Cursor, ledgerCloseTimestamp int64, txHash *xdr.Hash) bool {
+		cursorID := cursor.String()
+		if _, exists := cursorSet[cursorID]; exists {
+			return true
+		}
+
+		if request.Matches(event) && cursor.Cmp(start) >= 0 {
+			cursorSet[cursorID] = struct{}{}
+			found = append(found, entry{cursor, ledgerCloseTimestamp, event, txHash})
+		}
+		return uint(len(found)) < limit
+	}
+
+	err = h.dbReader.GetEvents(ctx, cursorRange, contractIDs, f)
```

Comment:
I am concerned about the efficiency of this. When events were stored in-memory it was probably fine to scan through a potentially large range of events and filter them by applying a function to all of them.

I suspect that the same approach can be prohibitive when moving the events to a database and extending the default history to 7 days.

It's good that we filter by contract ID on the database but to ensure this performs decently we may either need to either:
1. Be more restrictive (e.g. enforce providing contract IDs, with a cap on how many contracts to provide)
2. Move more of the filtering to the DB

Reply:
Regardless I would like to see a benchmark for event querying

Reply:
Ideally this should never scan 7 days of history as we are restricting fetch query by [this](https://github.com/stellar/soroban-rpc/pull/215/files#diff-1f027a1543c912ae4bc1533d9fece665a2b555548e8f74a8e591a70fdc317b2fR362)

Some context around Ledger restriction for given request here in design [doc](https://docs.google.com/document/d/1Q5AJvszwFmwl3NzHh7PJIMj4mhN3HgHRn_7IOQGeMWg/edit#heading=h.ah30ftbvhnx7)

Ledger restriction parameter (N) should be more explicit and ideally should be part of the config. so for any given get request we will only look at start: start + N ledger. Also it might be also useful to include lastProcessedLedger as a part of response so client understands what to use as startLedger in follow up query. I will add this change once we agree.

I also liked the idea of making filters more restrictive because current filters are more loose ( contract_ids, event_type, topics all are optional) but this can be a additive change along with ledger restriction part.